src/briefcase/platforms/web/static.py

```diff
@@ -0,0 +1,352 @@
+        with ZipFile(wheelfile) as wheel:
+            for filename in wheel.namelist():
+                path = Path(filename)
+                # Any CSS file in a `static` folder is appended
+                if (
+                    len(path.parts) > 1
+                    and path.parts[1] == "static"
+                    and path.suffix == ".css"
+                ):
+                    self.logger.info(f"Found {filename}")
+                    css_file.write(
+                        "\n/*******************************************************\n"
+                    )
+                    css_file.write(f" * {package}::{'/'.join(path.parts[2:])}\n")
+                    css_file.write(
+                        " *******************************************************/\n\n"
+                    )
+                    css_file.write(wheel.read(filename).decode("utf-8"))
+
+    def build_app(self, app: AppConfig, **kwargs):
+        """Build the static web deployment for the application.
+
+        :param app: The application to build
+        """
+        self.logger.info("Building web project...", prefix=app.app_name)
+
+        if self.wheel_path(app).exists():
+            with self.input.wait_bar("Removing old wheels..."):
+                self.tools.shutil.rmtree(self.wheel_path(app))
+
+        self.wheel_path(app).mkdir(parents=True)
+
+        with self.input.wait_bar("Building app wheel..."):
+            try:
+                self.tools.subprocess.run(
+                    [
+                        sys.executable,
+                        "-u",
+                        "-m",
+                        "wheel",
+                        "pack",
+                        self.app_path(app),
+                        "--dest-dir",
+                        self.wheel_path(app),
+                    ],
+                    check=True,
+                )
+            except subprocess.CalledProcessError as e:
+                raise BriefcaseCommandError(
+                    f"Unable to build wheel for app {app.app_name!r}"
+                ) from e
+
+        with self.input.wait_bar("Installing wheels for dependencies..."):
+            try:
+                self.tools.subprocess.run(
+                    [
+                        sys.executable,
+                        "-u",
+                        "-m",
+                        "pip",
+                        "wheel",
+                        "--wheel-dir",
+                        self.wheel_path(app),
+                        "-r",
+                        self.bundle_path(app) / "requirements.txt",
+                    ],
+                    check=True,
+                )
+            except subprocess.CalledProcessError as e:
+                raise BriefcaseCommandError(
+                    f"Unable to install dependencies for app {app.app_name!r}"
+                ) from e
+
+        with self.input.wait_bar("Writing Pyscript configuration file..."):
```

Comment:
disclosure: i consider this at least borderline nitpicking.

This section of logging is difficult for me to easily parse:
```
Writing Pyscript configuration file... done
Compile static web content from wheels...
Processing mywebapp-0.0.1-py3-none-any.whl...
Processing toga_core-0.3.0.dev39-py3-none-any.whl...
Processing toga_dummy-0.3.0.dev39-py3-none-any.whl...
Processing toga_web-0.3.0.dev39-py3-none-any.whl...
Found toga_web/static/toga.css
Processing travertino-0.1.3-py3-none-any.whl...
Compiling static web content from wheels... done
```
1) There are three important things happening here but it's hard to separate them out.
2) There are a lot of ellipsis....but only some say `done` after them.

Two possible improvements.

1) Consider the output from `pip` that directly proceeds this:
```
Processing /home/user/github/freakboy3742/toga/src/core
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Processing /home/user/github/freakboy3742/toga/src/dummy
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Processing /home/user/github/freakboy3742/toga/src/web
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Collecting travertino>=0.1.3
  Using cached travertino-0.1.3-py3-none-any.whl (15 kB)
Saved ./web/static/mywebapp/www/static/wheels/travertino-0.1.3-py3-none-any.whl
Building wheels for collected packages: toga-core, toga-dummy, toga-web
  Building wheel for toga-core (setup.py): started
  Building wheel for toga-core (setup.py): finished with status 'done'
...
```
This use of the indent makes it dramatically easier to quickly skim the output. Using an indent for the `Processing <wheel>` output really improves readability IMO:
```
Writing Pyscript configuration file... done
Compile static web content from wheels...
  Processing mywebapp-0.0.1-py3-none-any.whl...
  Processing toga_core-0.3.0.dev39-py3-none-any.whl...
  Processing toga_dummy-0.3.0.dev39-py3-none-any.whl...
  Processing toga_web-0.3.0.dev39-py3-none-any.whl...
Found toga_web/static/toga.css
  Processing travertino-0.1.3-py3-none-any.whl...
Compiling static web content from wheels... done
```
I guess if we really wanted to, we could add an `indent` argument to `Log` to manage this...

2) As for all the ellipsis...

I've been wondering for a while if ellipsis should be exclusively reserved for the Wait Bar. If the indent is added, then I definitely think it makes sense to at least avoid the ellipsis for the `Compile static web content from wheels` log.

As for the ellipsis for the `Processing <wheel>` log, I was initially thinking they should be Wait bars....but that would mean the Wait bar would be changing very quickly and that isn't great. So, probably best to get rid of the ellipsis there too and just let the single Wait Bar convey that tasks are ongoing.

This is where this would ultimately end up:
```
[mywebapp] Building web project...
Removing old wheels... done
Repacking wheel as /home/user/tmp/beeware/mywebapp/web/static/mywebapp/www/static/wheels/mywebapp-0.0.1-py3-none-any.whl...OK
Building app wheel... done
Processing /home/user/github/freakboy3742/toga/src/core
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Processing /home/user/github/freakboy3742/toga/src/dummy
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Processing /home/user/github/freakboy3742/toga/src/web
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Collecting travertino>=0.1.3
  Using cached travertino-0.1.3-py3-none-any.whl (15 kB)
Saved ./web/static/mywebapp/www/static/wheels/travertino-0.1.3-py3-none-any.whl
Building wheels for collected packages: toga-core, toga-dummy, toga-web
  Building wheel for toga-core (setup.py): started
  Building wheel for toga-core (setup.py): finished with status 'done'
  Created wheel for toga-core: filename=toga_core-0.3.0.dev39-py3-none-any.whl size=496221 sha256=7d63aaeae0a49ac0b40b5ee57541b8d4f3df935c3e8c5832dae548df32cd7cae
  Stored in directory: /tmp/pip-ephem-wheel-cache-ox82hom4/wheels/d6/c5/94/ed3559e7d1e227a880e8f68277a009ba9faa83254cdcb5fbce
  Building wheel for toga-dummy (setup.py): started
  Building wheel for toga-dummy (setup.py): finished with status 'done'
  Created wheel for toga-dummy: filename=toga_dummy-0.3.0.dev39-py3-none-any.whl size=27249 sha256=e28701baadd3879cc8530082f8be391c54982dbae3584db35257960c7484ae65
  Stored in directory: /tmp/pip-ephem-wheel-cache-ox82hom4/wheels/c0/a2/58/b2501a1be28389106869c459b744fec9e6f5631ada886e2e74
  Building wheel for toga-web (setup.py): started
  Building wheel for toga-web (setup.py): finished with status 'done'
  Created wheel for toga-web: filename=toga_web-0.3.0.dev39-py3-none-any.whl size=13068 sha256=d369f6eac3c796b1579e223f65f5c0efc17c332aaf868f1071281c979b580f53
  Stored in directory: /tmp/pip-ephem-wheel-cache-ox82hom4/wheels/19/e5/bf/617886587698943f953cf16ea7e3623985fedffbc8b187e98c
Successfully built toga-core toga-dummy toga-web
Installing wheels for dependencies... done
Writing Pyscript configuration file... done
Compile static web content from wheels
  Processing mywebapp-0.0.1-py3-none-any.whl
  Processing toga_core-0.3.0.dev39-py3-none-any.whl
  Processing toga_dummy-0.3.0.dev39-py3-none-any.whl
  Processing toga_web-0.3.0.dev39-py3-none-any.whl
  Found toga_web/static/toga.css
  Processing travertino-0.1.3-py3-none-any.whl
Compiling static web content from wheels... done
```

Note: `pip` has a whole infrastructure to manage the current level of indent and incrementing and decrementing it. I'd say limiting ourselves to a single indent level is reasonable.

Reply:
Completely agreed that the "wall of text" from that stage isn't great; I went through a couple of iterations of waitbar use before I settled where I did, and I still wasn't happy with the result.

I think the indentation and removing the ellipsis is a good improvement, though; I'll add those in.

More generally, I see 3 generic problems with the output as rendered in the build command:

Firstly, there's no vertical breaks. There's a lot being done, and being done quickly, but no vertical breaks to indicate "phases" in the build. We've been leaning away from empty calls to `info()`; but the stages are sufficiently short lived (and minor) that they don't really warrant a 'prefix' block. Not sure if there's a good solution here other than those two options.

Secondly, the need to have a "prefix" statement when using a waitbar that produces output. This is easy enough to do with a call to "info", but the "grammar" of those statements isn't clear. I broadly agree with your comment about keeping ellipsis exclusive to waitbars - an ellipsis should always be "closed" with "done" - but if you do that, the grammar of the "What I'm about to do" statement doesn't flow well to me. In this example, we need something to explain what "processing" means, but "Compile static web content" and "Compiling static web content" both feel like the wrong tense. "Compile" might make more sense if it was a prefixed marker, but as currently rendered, it feels like the best of a bad bunch of options, rather than a good option in itself.

Thirdly, there's the indentation issue. I agree indentation is a nice touch; although I do wonder if using indentation here is a crutch for getting over the first two issues. If we're going to start adopting indentation, then I agree having some sort of baked-in API would be a good idea - but the fact that we've been able to survive this long without indentation makes me wonder whether coming up with a good fixes for the first two might be the better (and easier) approach.

Reply:
Yeah...this is quite reminiscent of my ruminations in #755. I'm hesitant to go too much back to the drawing board given I didn't really find good answers then.

With that, I went through the output of the `package` command for all the platforms and formats before the apps were created so I could get a good sense of Briefcase output. I couldn't really identify anywhere that a lot of information generated by Briefcase is presented to the user. This may be the first case where particularly relevant information is being shown to the user and quickly. So, this may be the first use-case for something like indentation to increase readability.

In several of the outputs from tools, indentation use was certainly popular (especially with google).

I think indentation is a reasonable addition to logging. It would also be fairly trivial to add an "indent level" param to the APIs; this approach would also inherently limit how broadly this could be used since you wouldn't be able to call arbitrary functions that are "indent-unaware" (for lack of a better term).